File: library/src/main/java/me/proxer/library/api/DelimitedEnumSetAdapterFactory.java
package me.proxer.library.api;

import com.serjltt.moshi.adapters.FallbackEnum;
import com.squareup.moshi.Json;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.JsonDataException;
import com.squareup.moshi.JsonReader;
import com.squareup.moshi.JsonWriter;
import com.squareup.moshi.Moshi;
import com.squareup.moshi.Types;
import me.proxer.library.util.ProxerUtils;

import javax.annotation.Nullable;
import java.io.IOException;
import java.lang.annotation.Annotation;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.Arrays;
import java.util.Collections;
import java.util.EnumSet;
import java.util.Locale;
import java.util.Set;

/**
 * @author Ruben Gees
 */
class DelimitedEnumSetAdapterFactory implements JsonAdapter.Factory {

    @SuppressWarnings({"Duplicates", "unchecked"})
    @Nullable
    @Override
    public JsonAdapter<?> create(final Type type, final Set<? extends Annotation> annotations, final Moshi moshi) {
        final Type rawType = Types.getRawType(type);

        if (rawType != Set.class || !(type instanceof ParameterizedType)) {
            return null;
        }

        final Type parameterType = ((ParameterizedType) type).getActualTypeArguments()[0];

        if (!(parameterType instanceof Class<?>) || !((Class) parameterType).isEnum()) {
            return null;
        }

        final DelimitedEnumSet annotation = findAnnotation(annotations);

        if (annotation == null) {
            return null;
        }

        return new DelimitedEnumSetAdapter((Class<?>) parameterType, annotation.delimiter());
    }

    @Nullable
    private DelimitedEnumSet findAnnotation(final Set<? extends Annotation> annotations) {
        for (final Annotation annotation : annotations) {
            if (annotation.annotationType() == DelimitedEnumSet.class) {
                return (DelimitedEnumSet) annotation;
            }
        }

        return null;
    }

    private static class DelimitedEnumSetAdapter<T extends Enum<T>> extends JsonAdapter<Set<T>> {

        private final Class<T> enumType;
        private final String delimiter;

        private final T[] enumConstants;
        private final String[] enumNames;
        private final T fallbackEnum;

        DelimitedEnumSetAdapter(final Class<T> enumType, final String delimiter) {
            this.enumType = enumType;
            this.delimiter = delimiter;

            this.enumConstants = enumType.getEnumConstants();
            this.enumNames = new String[this.enumConstants.length];

            for (int i = 0; i < this.enumConstants.length; i++) {
                final T constant = this.enumConstants[i];

                try {
                    final Json annotation = enumType.getField(constant.name()).getAnnotation(Json.class);
                    final String name = annotation != null ? annotation.name() : constant.name();

                    this.enumNames[i] = name.toLowerCase(Locale.US);
                } catch (NoSuchFieldException e) {
                    throw new AssertionError("Missing field in " + enumType.getName(), e);
                }
            }

            final FallbackEnum fallbackAnnotation = enumType.getAnnotation(FallbackEnum.class);

            if (fallbackAnnotation != null) {
                fallbackEnum = Enum.valueOf(enumType, fallbackAnnotation.name());
            } else {
                fallbackEnum = null;
            }
        }

        @Override
        public Set<T> fromJson(final JsonReader reader) throws IOException {
            if (reader.peek() == JsonReader.Token.NULL) {
                reader.nextNull();

                return Collections.emptySet();
            } else {
                final String rawParts = reader.nextString().trim().toLowerCase(Locale.US);

                if (rawParts.isEmpty()) {
                    return Collections.emptySet();
                } else {
                    final String[] splitParts = rawParts.split(delimiter);

                    return convertToEnumConstants(reader, splitParts);
                }
            }
        }

        @Override
        public void toJson(final JsonWriter writer, @Nullable final Set<T> value) throws IOException {
            writer.value(value == null ? null : ProxerUtils.joinEnums(delimiter, EnumSet.copyOf(value)));
        }

        private Set<T> convertToEnumConstants(final JsonReader reader, final String... parts) {
            final EnumSet<T> result = EnumSet.noneOf(enumType);

            for (final String part : parts) {
                boolean found = false;

                for (int i = 0; i < enumNames.length; i++) {
                    final String name = enumNames[i];

                    if (part.equals(name)) {
                        result.add(enumConstants[i]);

                        found = true;
                        break;
                    }
                }

                if (!found) {
                    if (fallbackEnum != null) {
                        result.add(fallbackEnum);
                    } else {
                        throw new JsonDataException("Expected one of " + Arrays.asList(enumNames)
                                + " but was " + part + " at path " + reader.getPath());
                    }
                }
            }

            return result;
        }
    }
}

File: library/src/main/kotlin/me/proxer/library/internal/adapter/DelimitedEnumSetAdapterFactory.kt
package me.proxer.library.internal.adapter

import com.serjltt.moshi.adapters.FallbackEnum
import com.squareup.moshi.Json
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.JsonDataException
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import me.proxer.library.util.ProxerUtils
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type
import java.util.EnumSet
import java.util.Locale

/**
 * @author Ruben Gees
 */
internal class DelimitedEnumSetAdapterFactory : JsonAdapter.Factory {

    @Suppress("ReturnCount")
    override fun create(type: Type, annotations: Set<Annotation>, moshi: Moshi): JsonAdapter<*>? {
        val rawType = Types.getRawType(type)

        if (rawType !== Set::class.java || type !is ParameterizedType) {
            return null
        }

        val parameterType = type.actualTypeArguments[0]

        if (parameterType !is Class<*> || !parameterType.isEnum) {
            return null
        }

        val annotation = annotations.filterIsInstance(DelimitedEnumSet::class.java).firstOrNull() ?: return null

        // There is no other way in Kotlin to instantiate a generic class with an enum type parameter, when not
        // statically knowing the exact enum.
        // Idea from here: https://stackoverflow.com/a/46422600/4279995
        @Suppress("UNCHECKED_CAST")
        return DelimitedEnumSetAdapter(
            parameterType as Class<PlaceHolderEnum>,
            annotation.delimiter
        )
    }

    private class DelimitedEnumSetAdapter<T : Enum<T>> internal constructor(
        private val enumType: Class<T>,
        private val delimiter: String
    ) : JsonAdapter<Set<T>>() {

        private val enumMap = enumType.enumConstants
            .map { constant ->
                try {
                    val annotation = enumType.getField(constant.name).getAnnotation(Json::class.java)
                    val name = annotation?.name ?: constant.name

                    name.toLowerCase(Locale.US) to constant
                } catch (error: NoSuchFieldException) {
                    throw AssertionError("Missing field in ${enumType.name}", error)
                }
            }
            .toMap()

        private val fallbackEnum = enumType.getAnnotation(FallbackEnum::class.java)?.let {
            java.lang.Enum.valueOf(enumType, it.name)
        }

        override fun fromJson(reader: JsonReader): Set<T> {
            return if (reader.peek() == JsonReader.Token.NULL) {
                reader.nextNull<Any>()

                emptySet()
            } else {
                val rawParts = reader.nextString().trim().toLowerCase(Locale.US)

                if (rawParts.isEmpty()) {
                    emptySet()
                } else {
                    convertToEnumConstants(reader, rawParts.split(delimiter))
                }
            }
        }

        override fun toJson(writer: JsonWriter, value: Set<T>?) {
            if (value == null) {
                writer.nullValue()
            } else {
                writer.value(value.joinToString(delimiter) { ProxerUtils.getSafeApiEnumName(it) })
            }
        }

        private fun convertToEnumConstants(reader: JsonReader, parts: List<String>): Set<T> {
            return parts
                .map {
                    enumMap[it] ?: fallbackEnum ?: throw JsonDataException(
                        "Expected one of ${enumMap.keys} but was $it at path ${reader.path}"
                    )
                }
                .let { EnumSet.copyOf(it) }
        }
    }

    @Suppress("UnusedPrivateClass")
    private enum class PlaceHolderEnum
}
